File: pa2/Lex.c
//-----------------------------------------------------------------------------
// Name: Kelly Scanlon
// User: kpscanlo
// Assignment: pa2
//
// Lex.c
// 
//-----------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

#define MAX_LEN 160


char *strdup (const char *s) {
   char *d = malloc (strlen (s) + 1);
   if (d == NULL) return NULL;
   strcpy (d,s);
   return d;
}


int main (int argc, char * argv[]) {

   int n = 0;
   int lineNumber = 0;
   FILE *in, *out;
   char line[MAX_LEN];

   List L = newList();

   // check command line for correct number of arguments
   if( argc != 3 ){
     printf("Usage: %s <input file> <output file>\n", argv[0]);
     exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   if( in==NULL ){
     printf("Unable to open file %s for reading\n", argv[1]);
     exit(1);
   }
   while( fgets(line, MAX_LEN, in) != NULL ) {
      n++;
   }
   fclose(in);
   
   char* lines[n];

   in = fopen(argv[1], "r");
   while( fgets(line, MAX_LEN, in) != NULL) {
      lines[lineNumber] = strdup(line);
      if(lineNumber > 0) {
         moveTo(L, lineNumber-1);
         int i = lineNumber - 1;
         char* tmp;
         tmp = lines[lineNumber];
         while(i > 0 && strcmp(tmp, lines[getElement(L)])<0) {
            movePrev(L);
            i--;
         }
         if(strcmp(tmp, lines[getElement(L)])<0) 
            insertBefore(L, lineNumber);
         else
            insertAfter(L, lineNumber);
      } else {
         append(L, lineNumber);
      }
      lineNumber++;
   }
   fclose(in);

   out = fopen(argv[2], "w");
   if( out==NULL ){
     printf("Unable to open file %s for writing\n", argv[2]);
     exit(1);
   }

   moveTo(L, 0);
   for(int i = 0; i < n; i++) {
      fprintf(out, lines[getElement(L)]);
      moveNext(L);
   }
   
   for(int i = 0; i < n; i++) {
      free(lines[i]);
   }

   freeList(&L);
   fclose(out);

   return(0);

}

File: pa1/Lex.java
// Name: Kelly Scanlon
// User: kpscanlo
// Assignment: pa1

import java.io.*;
import java.util.Scanner;
import java.lang.String;


class Lex {
   public static void main(String[] args) throws IOException {
      Scanner in = null;
      PrintWriter out = null;
      int  n = 0, lineNumber = 0;
      List L = new List();
      String[] lines = null;

      if(args.length < 2) {
         System.err.println("Usage: FileIO infile outfile");
         System.exit(1);
      }

      in = new Scanner(new File(args[0]));
      while( in.hasNextLine() ) {
         n++;
         in.nextLine();
      }
      in.close();

      lines = new String[n];
      in = new Scanner(new File(args[0]));

      while( in.hasNextLine() ) {
         lines[lineNumber] = in.nextLine();
         if(lineNumber > 0) {
            L.moveTo(lineNumber - 1);
            int i = lineNumber - 1;
            String tmp = lines[lineNumber];
            while(i > 0 && tmp.compareTo(lines[L.getElement()])<0) {
               L.movePrev();
               i--;
            }
            if(tmp.compareTo(lines[L.getElement()])<0)
               L.insertBefore(lineNumber);
            else
               L.insertAfter(lineNumber);
         } else {
            L.append(lineNumber);
         }
         lineNumber++;
      }
      in.close();
      
      out = new PrintWriter(new FileWriter(args[1]));
      L.moveTo(0);
      for(int i = 0; i < n; i++) {
         out.println(lines[L.getElement()]);
         L.moveNext();
      }
      out.close();
   }
}

